src/components/widgetConfigs/timeseries/TimeseriesConfigurator.tsx

```diff
@@ -0,0 +1,220 @@
+// Copyright 2020 Cognite AS
+import React, { useState } from 'react';
+import { useForm, Controller } from 'react-hook-form';
+import Button from '@material-ui/core/Button';
+import TextField from '@material-ui/core/TextField';
+import CircularProgress from '@material-ui/core/CircularProgress';
+
+import Paper from '@material-ui/core/Paper';
+import Grid from '@material-ui/core/Grid';
+import './TimeseriesConfigurator.css';
+import { RootAction } from 'StoreTypes';
+import { bindActionCreators, Dispatch } from 'redux';
+
+import { setAlerts } from 'store/actions/root-action';
+import {
+  FormValues,
+  DefaultProps,
+  QueryParams,
+  ValueMapping,
+} from './interfaces';
+import { fetchTimeseries } from './timeseriesFetcher';
+import { getDynamicFields, timeUnitMapping } from './constants';
+
+/**
+ * Base component for timeseries related widget configurators.
+ */
+export const TimeseriesConfigurator = (props: TimeseriesConfiguratorProps) => {
+  const { onCreate, configFields } = props;
+  const [timeseriesLoading, setTimeseriesLoading] = useState(false);
+
+  const defaultValues: FormValues = {
+    timeseriesExternalId: '',
+    timeseriesName: '',
+    timeseriesUnit: '',
+    startValue: 2,
+    startUnit: 'd',
+    granularityValue: 4,
+    granularityUnit: 'd',
+    elapsedTimeEnabler: false,
+    timeDisplayKey: '',
+    maxPrecentageValue: 1,
+  };
+
+  const { handleSubmit, reset, control, getValues, setValue } = useForm({
+    defaultValues,
+  });
+
+  /**
+   * This activates when the Timeseries external id is entered and user leaves the text box.
+   * This passes user entered timeseries external id and fetch timeseries data.
+   * Timeseries name and unit is set in the form.
+   * If there is no timeseries for the entered timeseries external id, it will show an alert.
+   */
+  const onTsExternalIdBlur = async () => {
+    if (getValues().timeseriesExternalId) {
+      setTimeseriesLoading(true);
+      const timeseries = await fetchTimeseries(
+        getValues().timeseriesExternalId
+      );
+
+      if (timeseries) {
+        setValue('timeseriesName', timeseries.name);
+        setValue('timeseriesUnit', timeseries.unit);
+      } else {
+        setValue('timeseriesName', '');
+        setValue('timeseriesUnit', '');
+        props.setAlerts({
+          type: 'warning',
+          text: 'Timeseries not found!',
+          hideApp: false,
+        });
+      }
+      setTimeseriesLoading(false);
+    }
+  };
+
+  /**
+   *
+   * This activated when user submit the form.
+   * Return value is updated based on the required fields.
+   */
+  const onSubmit = (data: FormValues) => {
+    const queryParams: QueryParams = {
```

Comment:
Do you have to handle the situation where things such as timeseriesExternalId is not yet set?

Reply:
if timeseriesExternalId is not set, then this method will never fire